core/ir/instr_shared.c

```diff
@@ -2999,6 +3000,172 @@ instr_uses_fp_reg(instr_t *instr)
     return false;
 }
 
+instr_t *
+instr_convert_to_isa_regdeps(void *drcontext, instr_t *instr_from)
+{
+    /* Create instruction we convert instr_from into.
+     */
+    instr_t *instr_to = instr_create(drcontext);
+
+    /* Retrieve number of register destination operands from real ISA instruction.
+     * Note that a destination operand that is a memory renference should have its
+     * registers (if any) counted as source operands, since they are being read.
+     * We use [src|dst]_reg_used to keep track of registers we've seen and avoid
+     * duplicates.
+     * We use max_reg_size to keep track of the size of the largest containing register.
+     * max_reg_size is a uint instead of a opnd_size_t to avoid relying on OPSZ_ enum
+     * values.
+     * We convert max_reg_size to OPSZ_ and use it as the size for all register operands
+     * when we set the operands in the converted instruction later on.
+     */
+    bool src_reg_used[MAX_NUM_REGS];
+    memset((void *)src_reg_used, 0, sizeof(src_reg_used));
+    uint num_srcs = 0;
+    bool dst_reg_used[MAX_NUM_REGS];
+    memset((void *)dst_reg_used, 0, sizeof(dst_reg_used));
+    uint num_dsts = 0;
+    uint original_num_dsts = (uint)instr_num_dsts(instr_from);
+    uint max_reg_size = 0;
+    for (uint dst_index = 0; dst_index < original_num_dsts; ++dst_index) {
+        opnd_t dst_opnd = instr_get_dst(instr_from, dst_index);
+        uint num_regs_used_by_opnd = (uint)opnd_num_regs_used(dst_opnd);
+        if (opnd_is_memory_reference(dst_opnd)) {
+            for (uint opnd_index = 0; opnd_index < num_regs_used_by_opnd; ++opnd_index) {
+                reg_id_t reg = opnd_get_reg_used(dst_opnd, opnd_index);
+                /* Map sub-registers to their containing register.
+                 */
+                reg_id_t reg_canonical = reg_to_pointer_sized(reg);
+                opnd_size_t reg_size = reg_get_size(reg_canonical);
```

Comment:
I'm a bit confused.
I thought this was part of the obfuscation (through information loss) we wanted to introduce.
If we have instr A that operates on rax and rbx, and instr B on eax and ebx (in the real ISA) don't we want both to map to an instr C that operates on rax and rbx with operation size = 8 bytes (in the REG_DEPS ISA)?
Otherwise, if we use the actual register size (instead of the containing register size), we would have instr A that maps to an instr D (in REG_DEPS ISA) that operates on rax and rbx with operation size = 8 bytes, and instr B that maps to an instr E (again, in REG_DEPS ISA) that still operates on rax and rbx but with operation size = 4 bytes (which makes it pretty straightforward that instr E is actually operating on eax and ebx).
At that point, I think that the mapping of sub-registers to their containing registers does not hide much, might just as well use sub-registers ids?

Also, I believe storing the operation size is still necessary.
We can't extract register size info from a REGDEPS register like we can for real registers.
Their identifiers are just numbers, calling, for example, `reg_get_size()` on a REGDEPS register wouldn't return the correct answer because they are not "real registers", they just so happen to have a value in the range of DR_REG enum (`opnd_get_size()` is what should be used).

Reply:
As discussed offline, we now take the max source operand size (regardless of operands being registers or anything else).
Done.